File: src/admin/adapters/konnect.adapter.ts
import {
  createKonnectClient,
  type KonnectClient,
} from "@pymthouse/konnect-metering";
import type { components } from "@pymthouse/konnect-metering";
import { BackendUnreachableError } from "../errors.js";
import type { OpenMeterAdmin } from "../port.js";
import type {
  AdminCapabilities,
  Customer,
  CustomerInput,
  CustomerListQuery,
  Feature,
  FeatureInput,
  Meter,
  MeterInput,
  Plan,
  PlanInput,
  Subscription,
  SubscriptionInput,
} from "../types.js";

type ApiMeter = components["schemas"]["Meter"];
type ApiFeature = components["schemas"]["Feature"];
type ApiPlan = components["schemas"]["BillingPlan"];
type ApiCustomer = components["schemas"]["BillingCustomer"];
type ApiSubscription = components["schemas"]["BillingSubscription"];

function mapMeter(m: ApiMeter): Meter {
  return {
    id: m.id,
    key: m.key,
    name: m.name,
    dimensions: m.dimensions,
  };
}

function mapFeature(m: ApiFeature): Feature {
  return {
    id: m.id,
    key: m.key,
    name: m.name,
    meterId: m.meter?.id,
  };
}

function mapPlan(m: ApiPlan): Plan {
  return {
    id: m.id,
    key: m.key,
    name: m.name,
    status: m.status,
  };
}

function mapCustomer(m: ApiCustomer): Customer {
  return {
    id: m.id,
    key: m.key ?? m.id,
    name: m.name,
    subjectKeys: m.usage_attribution?.subject_keys,
  };
}

function mapSubscription(
  m: ApiSubscription,
  planKey?: string,
  customerId?: string,
): Subscription {
  return {
    id: m.id,
    status: m.status,
    planKey,
    customerId: customerId ?? m.customer_id,
  };
}

function buildUsageRateCard(input: {
  key: string;
  name: string;
  featureId: string;
  unitAmount: string;
  includedMicros: number;
}): components["schemas"]["BillingRateCard"] {
  const card: components["schemas"]["BillingRateCard"] = {
    key: input.key,
    name: input.name,
    feature: { id: input.featureId },
    billing_cadence: "P1M",
    payment_term: "in_arrears",
    price: {
      type: "unit",
      amount: input.unitAmount,
    },
  };

  if (input.includedMicros > 0) {
    card.discounts = {
      usage: String(input.includedMicros),
    };
  }

  return card;
}

export class KonnectAdmin implements OpenMeterAdmin {
  readonly capabilities: AdminCapabilities = {
    plans: "full",
    subscriptions: "full",
  };

  private readonly client: KonnectClient;

  constructor(opts: { baseUrl: string; apiKey: string }) {
    this.client = createKonnectClient(opts);
  }

  async waitForHealthy(): Promise<void> {
    try {
      await this.client.waitForHealthy();
    } catch (err) {
      throw new BackendUnreachableError(
        err instanceof Error ? err.message : String(err),
      );
    }
  }

  async listMeters(): Promise<Meter[]> {
    const { data } = await this.client.api.GET("/openmeter/meters");
    return (data?.data ?? []).map(mapMeter);
  }

  async createMeter(input: MeterInput): Promise<Meter> {
    const { data } = await this.client.api.POST("/openmeter/meters", {
      body: {
        key: input.key,
        name: input.name,
        description: input.description,
        event_type: input.eventType,
        aggregation: input.aggregation,
        value_property: input.valueProperty,
        dimensions: input.dimensions,
      },
    });
    if (!data) {
      throw new Error(`Failed to create meter ${input.key}`);
    }
    return mapMeter(data);
  }

  async listFeatures(): Promise<Feature[]> {
    const { data } = await this.client.api.GET("/openmeter/features");
    return (data?.data ?? []).map(mapFeature);
  }

  async createFeature(input: FeatureInput): Promise<Feature> {
    const { data } = await this.client.api.POST("/openmeter/features", {
      body: {
        key: input.key,
        name: input.name,
        meter: { id: input.meterId },
      },
    });
    if (!data) {
      throw new Error(`Failed to create feature ${input.key}`);
    }
    return mapFeature(data);
  }

  async listPlans(): Promise<Plan[]> {
    const { data } = await this.client.api.GET("/openmeter/plans");
    return (data?.data ?? []).map(mapPlan);
  }

  async ensurePlan(input: PlanInput): Promise<Plan> {
    const includedMicros = Math.max(0, Math.floor(input.includedMicros));

    const features = await this.listFeatures();
    let featureId = features.find((f) => f.key === input.featureKey)?.id;

    if (!featureId) {
      const created = await this.createFeature({
        key: input.featureKey,
        name: input.featureName,
        meterId: input.billableMeterId,
      });
      featureId = created.id;
    }

    const rateCard = buildUsageRateCard({
      key: input.featureKey,
      name: "Billable usage",
      featureId,
      unitAmount: input.unitAmount,
      includedMicros,
    });

    const { data: createdPlan } = await this.client.api.POST("/openmeter/plans", {
      body: {
        key: input.key,
        name: input.name,
        currency: input.currency ?? "USD",
        billing_cadence: input.billingCadence ?? "P1M",
        pro_rating_enabled: true,
        phases: [
          {
            key: "default",
            name: "Default",
            rate_cards: [rateCard],
          },
        ],
      },
    });

    if (!createdPlan) {
      throw new Error(`Failed to create plan ${input.key}`);
    }

    if (createdPlan.status === "draft") {
      await this.client.api.POST("/openmeter/plans/{planId}/publish", {
        params: { path: { planId: createdPlan.id } },
      });
    }

    return {
      id: createdPlan.id,
      key: input.key,
      name: input.name,
    };
  }

  async listCustomers(query?: CustomerListQuery): Promise<Customer[]> {
    if (query?.key) {
      const { data } = await this.client.api.GET("/openmeter/customers", {
        params: {
          query: {
            filter: { key: query.key },
            page: { number: 1, size: 100 },
          },
        },
      });
      const rows = data?.data ?? [];
      return rows.filter((row) => row.key === query.key).map(mapCustomer);
    }

    const { data } = await this.client.api.GET("/openmeter/customers", {
      params: {
        query: {
          page: {
            number: query?.page ?? 1,
            size: query?.pageSize ?? 100,
          },
        },
      },
    });
    return (data?.data ?? []).map(mapCustomer);
  }

  async createCustomer(input: CustomerInput): Promise<Customer> {
    const { data } = await this.client.api.POST("/openmeter/customers", {
      body: {
        key: input.key,
        name: input.name,
        usage_attribution: { subject_keys: input.subjectKeys },
      },
    });
    if (!data) {
      throw new Error(`Failed to create customer ${input.key}`);
    }
    return mapCustomer(data);
  }

  async listCustomerSubscriptions(customerId: string): Promise<Subscription[]> {
    const { data } = await this.client.api.GET("/openmeter/subscriptions", {
      params: {
        query: {
          filter: { customer_id: customerId },
        },
      },
    });
    const subs = data?.data ?? [];
    const plans = subs.some((sub) => sub.plan_id)
      ? await this.listPlans()
      : [];
    const planKeyById = new Map(plans.map((plan) => [plan.id, plan.key]));

    return subs.map((sub) =>
      mapSubscription(
        sub,
        sub.plan_id ? planKeyById.get(sub.plan_id) : undefined,
        customerId,
      ),
    );
  }

  async createSubscription(input: SubscriptionInput): Promise<Subscription> {
    const { data } = await this.client.api.POST("/openmeter/subscriptions", {
      body: {
        customer: { id: input.customerId },
        plan: { key: input.planKey },
      },
    });
    if (!data) {
      throw new Error(
        `Failed to create subscription for plan ${input.planKey}`,
      );
    }
    return mapSubscription(data, input.planKey, input.customerId);
  }
}

File: internal/admin/konnect_adapter.go
package admin

import (
	"context"
	"fmt"
	"time"

	sdkkonnectgo "github.com/Kong/sdk-konnect-go"
	"github.com/Kong/sdk-konnect-go/models/components"
	"github.com/Kong/sdk-konnect-go/models/operations"
)

type KonnectAdmin struct {
	sdk *sdkkonnectgo.SDK
}

type KonnectAdminConfig struct {
	APIKey  string
	BaseURL string
}

func NewKonnectAdmin(cfg KonnectAdminConfig) *KonnectAdmin {
	opts := []sdkkonnectgo.SDKOption{
		sdkkonnectgo.WithSecurity(components.Security{
			PersonalAccessToken: sdkkonnectgo.Pointer(cfg.APIKey),
		}),
	}
	if cfg.BaseURL != "" {
		opts = append(opts, sdkkonnectgo.WithServerURL(cfg.BaseURL))
	} else {
		opts = append(opts, sdkkonnectgo.WithServerIndex(1)) // us.api.konghq.com
	}
	return &KonnectAdmin{sdk: sdkkonnectgo.New(opts...)}
}

func NewKonnectAdminWithSDK(sdk *sdkkonnectgo.SDK) *KonnectAdmin {
	return &KonnectAdmin{sdk: sdk}
}

func (k *KonnectAdmin) WaitForHealthy(ctx context.Context) error {
	const maxAttempts = 15
	const delay = 2 * time.Second

	for i := range maxAttempts {
		res, err := k.sdk.Meters.ListMeters(ctx, operations.ListMetersRequest{
			Page: &components.PagePaginationQuery{
				Number: sdkkonnectgo.Pointer(int64(1)),
				Size:   sdkkonnectgo.Pointer(int64(1)),
			},
		})
		if err == nil && res.StatusCode >= 200 && res.StatusCode < 300 {
			return nil
		}
		if i < maxAttempts-1 {
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(delay):
			}
		}
	}
	return NewBackendUnreachableError("Konnect Metering & Billing not reachable")
}

func (k *KonnectAdmin) ListMeters(ctx context.Context) ([]Meter, error) {
	res, err := k.sdk.Meters.ListMeters(ctx, operations.ListMetersRequest{})
	if err != nil {
		return nil, fmt.Errorf("listing meters: %w", err)
	}
	var out []Meter
	if res.MeterPagePaginatedResponse != nil {
		for _, m := range res.MeterPagePaginatedResponse.Data {
			out = append(out, Meter{
				ID:         m.ID,
				Key:        m.Key,
				Name:       m.Name,
				Dimensions: m.Dimensions,
			})
		}
	}
	return out, nil
}

func (k *KonnectAdmin) CreateMeter(ctx context.Context, input MeterInput) (*Meter, error) {
	req := components.CreateMeterRequest{
		Key:           input.Key,
		Name:          input.Name,
		Description:   sdkkonnectgo.Pointer(input.Description),
		EventType:     input.EventType,
		Aggregation:   components.Aggregation(input.Aggregation),
		ValueProperty: nilIfEmpty(input.ValueProperty),
		Dimensions:    input.Dimensions,
	}
	res, err := k.sdk.Meters.CreateMeter(ctx, req)
	if err != nil {
		return nil, fmt.Errorf("creating meter %s: %w", input.Key, err)
	}
	if res.Meter == nil {
		return nil, fmt.Errorf("creating meter %s: empty response", input.Key)
	}
	return &Meter{
		ID:         res.Meter.ID,
		Key:        res.Meter.Key,
		Name:       res.Meter.Name,
		Dimensions: res.Meter.Dimensions,
	}, nil
}

func (k *KonnectAdmin) ListFeatures(ctx context.Context) ([]Feature, error) {
	res, err := k.sdk.OpenMeterFeatures.ListFeatures(ctx, operations.ListFeaturesRequest{})
	if err != nil {
		return nil, fmt.Errorf("listing features: %w", err)
	}
	var out []Feature
	if res.FeaturePagePaginatedResponse != nil {
		for _, f := range res.FeaturePagePaginatedResponse.Data {
			feat := Feature{ID: f.ID, Key: f.Key, Name: f.Name}
			if f.Meter != nil {
				feat.MeterID = f.Meter.ID
			}
			out = append(out, feat)
		}
	}
	return out, nil
}

func (k *KonnectAdmin) CreateFeature(ctx context.Context, input FeatureInput) (*Feature, error) {
	req := components.CreateFeatureRequest{
		Key:  input.Key,
		Name: input.Name,
	}
	if input.MeterID != "" {
		req.Meter = &components.CreateFeatureRequestMeterReference{ID: input.MeterID}
	}
	res, err := k.sdk.OpenMeterFeatures.CreateFeature(ctx, req)
	if err != nil {
		return nil, fmt.Errorf("creating feature %s: %w", input.Key, err)
	}
	if res.Feature == nil {
		return nil, fmt.Errorf("creating feature %s: empty response", input.Key)
	}
	feat := &Feature{ID: res.Feature.ID, Key: res.Feature.Key, Name: res.Feature.Name}
	if res.Feature.Meter != nil {
		feat.MeterID = res.Feature.Meter.ID
	}
	return feat, nil
}

func (k *KonnectAdmin) ListPlans(ctx context.Context) ([]Plan, error) {
	res, err := k.sdk.OpenMeterProductCatalog.ListPlans(ctx, operations.ListPlansRequest{})
	if err != nil {
		return nil, fmt.Errorf("listing plans: %w", err)
	}
	var out []Plan
	if res.PlanPagePaginatedResponse != nil {
		for _, p := range res.PlanPagePaginatedResponse.Data {
			out = append(out, Plan{
				ID:     p.ID,
				Key:    p.Key,
				Name:   p.Name,
				Status: string(p.Status),
			})
		}
	}
	return out, nil
}

func (k *KonnectAdmin) EnsurePlan(ctx context.Context, input PlanInput) (*Plan, error) {
	// Resolve feature ID
	features, err := k.ListFeatures(ctx)
	if err != nil {
		return nil, err
	}
	var featureID string
	for _, f := range features {
		if f.Key == input.FeatureKey {
			featureID = f.ID
			break
		}
	}
	if featureID == "" {
		created, err := k.CreateFeature(ctx, FeatureInput{
			Key:     input.FeatureKey,
			Name:    input.FeatureName,
			MeterID: input.BillableMetterID,
		})
		if err != nil {
			return nil, err
		}
		featureID = created.ID
	}

	// Pure pay-per-use rate card: no included units / discounts. Trial credit is
	// granted once per customer as an entitlement at provision time, not baked
	// into the plan (a plan-level discount recurs every billing period).
	rateCard := components.BillingRateCard{
		Key:            input.FeatureKey,
		Name:           "Billable usage",
		Feature:        &components.FeatureReference{ID: featureID},
		BillingCadence: sdkkonnectgo.Pointer("P1M"),
		PaymentTerm:    components.PaymentTermInArrears.ToPointer(),
		Price: components.CreatePriceUnit(components.BillingPriceUnit{
			Amount: input.UnitAmount,
		}),
	}

	currency := input.Currency
	if currency == "" {
		currency = "USD"
	}
	cadence := input.BillingCadence
	if cadence == "" {
		cadence = "P1M"
	}

	createRes, err := k.sdk.OpenMeterProductCatalog.CreatePlan(ctx, components.CreatePlanRequest{
		Key:              input.Key,
		Name:             input.Name,
		Currency:         currency,
		BillingCadence:   cadence,
		ProRatingEnabled: sdkkonnectgo.Pointer(true),
		Phases: []components.BillingPlanPhase{
			{
				Key:       "default",
				Name:      "Default",
				RateCards: []components.BillingRateCard{rateCard},
			},
		},
	})
	if err != nil {
		return nil, fmt.Errorf("creating plan %s: %w", input.Key, err)
	}
	if createRes.BillingPlan == nil {
		return nil, fmt.Errorf("creating plan %s: empty response", input.Key)
	}

	plan := createRes.BillingPlan
	if plan.Status == components.BillingPlanStatusDraft {
		_, err := k.sdk.OpenMeterProductCatalog.PublishPlan(ctx, plan.ID)
		if err != nil {
			return nil, fmt.Errorf("publishing plan %s: %w", input.Key, err)
		}
	}

	return &Plan{
		ID:   plan.ID,
		Key:  input.Key,
		Name: input.Name,
	}, nil
}

func (k *KonnectAdmin) DeleteMeter(ctx context.Context, id string) error {
	_, err := k.sdk.Meters.DeleteMeter(ctx, id)
	if err != nil {
		return fmt.Errorf("deleting meter %s: %w", id, err)
	}
	return nil
}

func (k *KonnectAdmin) DeleteFeature(ctx context.Context, id string) error {
	_, err := k.sdk.OpenMeterFeatures.DeleteFeature(ctx, id)
	if err != nil {
		return fmt.Errorf("deleting feature %s: %w", id, err)
	}
	return nil
}

func (k *KonnectAdmin) DeletePlan(ctx context.Context, id string) error {
	_, err := k.sdk.OpenMeterProductCatalog.DeletePlan(ctx, id)
	if err != nil {
		_, archiveErr := k.sdk.OpenMeterProductCatalog.ArchivePlan(ctx, id)
		if archiveErr != nil {
			return fmt.Errorf("deleting plan %s: %w (archive: %v)", id, err, archiveErr)
		}
	}
	return nil
}

func nilIfEmpty(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}
